src/suggestions.ts

```diff
@@ -0,0 +1,120 @@
+ * @param source - Path to a directory somewhere in a Git repository worktree
+ * @param violationSpecs - Specifiers for violations to repair
+ * @returns Fulfills to an array of patch suggestions
+ */
+export async function generatePatchSuggestions(
+  soraldJar: PathLike,
+  source: PathLike,
+  violationSpecs: string[]
+): Promise<PatchSuggestion[]> {
+  const repo = new Repo(source);
+  const worktreeRoot = await repo.getWorktreeRoot();
+  const throwawayStatsFile = '__sorald_throwaway_stats_file.json';
+
+  const allSuggestions: PatchSuggestion[] = [];
+  for (const spec of violationSpecs) {
+    await sorald.repair(soraldJar, source, throwawayStatsFile, [spec]);
+    const diff = await repo.diff();
+    const hunks = git.parseDiffHunks(diff);
+    for (const hunk of hunks) {
+      const suggestion = await generatePatchSuggestion(
+        hunk,
+        spec,
+        path.join(worktreeRoot.toString(), hunk.leftFile.toString())
+      );
+      allSuggestions.push(suggestion);
+    }
+    await repo.restore();
+  }
+  return allSuggestions;
+}
+
+async function generatePatchSuggestion(
+  hunk: Hunk,
+  spec: string,
+  localFile: PathLike
+): Promise<PatchSuggestion> {
+  const suggestionAdditions = hunk.additions.join('\n');
+  const suggestion =
+    hunk.leftRange.start === hunk.leftRange.end
+      ? `${await readLine(
+          localFile,
+          hunk.leftRange.start
+        )}\n${suggestionAdditions}`
+      : suggestionAdditions;
+  return {
+    linesToReplace: hunk.leftRange,
+    file: hunk.leftFile,
+    suggestion: `\`\`\`suggestion
+${suggestion}
+\`\`\``,
+    violationSpec: spec
+  };
+}
+
+async function readLine(filepath: PathLike, line: number): Promise<string> {
+  const content = await fs.promises.readFile(filepath, {encoding: 'utf8'});
+  return content.toString().split('\n')[line - 1];
+}
+
+/**
+ * Post a pach suggestion to the current pull request, assuming the GitHub
+ * context is in fact a pull request.
+ *
+ * @param ps - A patch suggestion to post
+ * @param token - Token to authenticate with the GitHub API
+ */
+export async function postPatchSuggestion(
+  ps: PatchSuggestion,
+  token: string
+): Promise<void> {
+  const octokit = github.getOctokit(token);
+  const pull_request = github.context.payload.pull_request;
+
+  if (pull_request !== undefined) {
+    const startLine = ps.linesToReplace.start;
+    const endLine = ps.linesToReplace.end - 1;
+
+    const lineArgs =
+      endLine <= startLine
+        ? {line: startLine}
+        : {start_line: startLine, line: endLine};
+
+    await octokit.rest.pulls.createReviewComment({
+      ...github.context.repo,
+      ...lineArgs,
+      commit_id: pull_request.head.sha,
+      pull_number: pull_request.number,
+      body: `To fix violation '${ps.violationSpec}', Sorald suggests the following:
+${ps.suggestion}
+`,
+      path: ps.file.toString(),
+      start_side: 'RIGHT'
+    });
+  }
+}
```

Comment:
This function is only tested manually right now, see spoonlabs/sorald#534. Will try to set up a system test at some later time.

Reply:
Yes, I figured that out. You will have to actually send a post request to GitHub for that and check the response to test it. Will that be feasible?

Reply:
For a full-on system test that actually validates the expected behavior: yes.

But I can also redesign the function to be more easily testable, and test everything but the actual post. For example, I can have it take a callback function and assert that said callback is called with the correct arguments. Or I can completely redesign the flow of operations, such that this function just creates the object to be posted, rather than actually perform the post.

As this GitHub Action is mostly to serve as a prototype for an eventual bot, I'll probably go with testing everything but the post itself.